### PR TITLE
[DataFrame] Implement loc, iloc

### PR DESCRIPTION
**Pull request by simon-mo**
## What do these changes do?

This PR implements basic `loc` and `iloc` for the dataframe. 

Out of five allowed inputs, this PR implements the most commonly used three:
- [x] A single label, e.g. 5 or 'a', (note that 5 is interpreted as a label of the index, and never as an integer position along the index).
- [x] A list or array of labels, e.g. ['a', 'b', 'c'].
- [x] A slice object with labels, e.g. 'a':'f' (note that contrary to usual python slices, both the start and the stop are included!).
- ~~A boolean array.~~
- ~~A callable function with one argument (the calling Series, DataFrame or Panel) and that returns valid output for indexing (one of the above)~~

### CODE REVIEW

**Comment by simon-mo** [comment]
This PR depends on commit 9b69cfc in PR #1599

File: python/ray/dataframe/indexing.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        if indexer == 'loc':
+            def retrieve_func(
+                df, idx_lst, col_label): return df.loc[idx_lst, col_label]
+        elif indexer == 'iloc':
+            def retrieve_func(
+                df, idx_lst, col_idx): return df.iloc[idx_lst, col_idx]
+
+        retrieved_rows_remote = []
+        for partition, idx_to_lookup in lookup_dict.items():
```
Comment (devin-petersohn) [inline python/ray/dataframe/indexing.py]: I prefer to wrap this in list comprehension:
```python
retrieved_rows_remote = [_deploy_func.remote(retrieve_func, self.df._df[partition], idx_to_lookup, col_list) for partition, idx_to_lookup in lookup_dict.items()]
```

File: python/ray/dataframe/indexing.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import pandas as pd
+import ray
+from .dataframe import _deploy_func
+
+
+def _row_tuples_to_dict(row_tuples):
```
Comment (devin-petersohn) [inline python/ray/dataframe/indexing.py]: Would it be faster to do a `self._index.groupby('partition')`?

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
A couple of minor changes. Looking good!

**Review by devin-petersohn — APPROVED** [review]
This is great! Thanks @simon-mo !

**Pull request merged**